### DISCUSSION TITLE
ESP partition format

### DISCUSSION THREAD

**Discussion by ragazenta** [systemd/mkosi#2642] [Q&A] [answered]
Hi, does `mkosi` or perhaps `systemd-repart` make some additional formatting to ESP partition?

We've been trying to build Raspberry Pi OS image using `mkosi`. It works with many workarounds.

Raspberry Pi requires a FAT partition to put firmware files and kernel, so I create custom repart definitions in mkosi.repart/

```ini
[Partition]
Type=ebd0a0a2-b9e5-4433-87c0-68b6b72699c7
Label=bootfs
Format=vfat
CopyFiles=/boot/firmware:/
SizeMinBytes=512M
SizeMaxBytes=512M
```
```ini
[Partition]
Type=root
Label=rootfs
Format=ext4
CopyFiles=/
ExcludeFiles=/boot/firmware
Minimize=guess
```

With above definitions, OS can boot fine. But if change the bootfs to use `Type=esp`, the result is somehow detected as FAT-16.

`Type=ebd0a0a2-b9e5-4433-87c0-68b6b72699c7` is UUID for Microsoft Basic Data Partition. If I manually change partition type code using `gdisk`, from `0700` -> `EF00`, OS also boot fine. So I think there is something when `mkosi` create disk partition.

After skimming systemd-repart code, it seems `Format=vfat` is hard coded to use FAT-32. Is there anything I miss?

**Comment by daandemeyer** [comment, marked as answer]
See the following code:

```
    partitions = make_disk(context, skip=("esp", "xbootldr"), tabs=True, msg="Generating disk image")
    install_kernel(context, partitions)
    normalize_mtime(context.root, context.config.source_date_epoch, directory=Path("boot"))
    normalize_mtime(context.root, context.config.source_date_epoch, directory=Path("efi"))
    partitions = make_disk(context, msg="Formatting ESP/XBOOTLDR partitions")
```

Without `Type=esp` both partitions will be created before we install the kernel to /boot, with `Type=esp`, the ESP partition will only be created after we install the kernel to /boot. I assume that's where the issue is coming from.

**Reply by ragazenta** [reply]
I see, thanks.

On a second thought, it doesn't feel right to use `esp` for Raspberry Pi when the bootloader is not uefi.

**Comment by septatrix** [comment]
> We've been trying to build Raspberry Pi OS image using `mkosi`. It works with many workarounds.

Do you mean you recreated [RPi-Distro/pi-gen](https://github.com/RPi-Distro/pi-gen) but replaced the build process with mkosi? If that is the case would it be possible to get access to the repo? I also tried that a while ago (#2103) but never really got to try it out due to a lack of time.

**Reply by ragazenta** [reply]
Yes, but not really equal to pi-gen. I removed some stuffs.

You could see the repo [here](https://github.com/ragazenta/mkpiosi). I've just created it, if you're interested.
